src/ngx_http_graphite_module.c

```diff
@@ -1794,14 +1813,111 @@ ngx_http_graphite_timer_event_handler(ngx_event_t *ev) {
         char *next = NULL;
         char *nl = NULL;
 
-        int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+        int fd;
+         if (ngx_strcmp(gmcf->protocol.data, "tcp") == 0) {
+             fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+         } else {
+             fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+         }
+
         if (fd < 0) {
             ngx_log_error(NGX_LOG_ALERT, ev->log, 0, "graphite can't create socket");
             if (!(ngx_quit || ngx_terminate || ngx_exiting))
                 ngx_add_timer(ev, gmcf->frequency);
             return;
         }
 
+        if (ngx_strcmp(gmcf->protocol.data, "tcp") == 0) {
+
+            int flags;
+            fd_set myset;
+            struct timeval tv;
+            int valopt;
+            socklen_t lon;
+            /* Set socket to non-blocking */
+            if ((flags = fcntl(fd, F_GETFL, 0)) < 0)
+            {
+                ngx_log_error(NGX_LOG_ALERT, ev->log, 0, "failed to get the current non-blocking flag");
+                if (!(ngx_quit || ngx_terminate || ngx_exiting))
+                    ngx_add_timer(ev, gmcf->frequency);
+                return;
+            }
+
+            if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0)
+            {
+                ngx_log_error(NGX_LOG_ALERT, ev->log, 0, "failed to set the socket non-blocking");
+                if (!(ngx_quit || ngx_terminate || ngx_exiting))
+                    ngx_add_timer(ev, gmcf->frequency);
+                return;
+            }
+              // Trying to connect with timeout
+              int res = connect(fd, (struct sockaddr *)&sin, sizeof(sin));
+              if (res < 0) {
+                 if (errno == EINPROGRESS) {
+                    ngx_log_error(NGX_LOG_ALERT, ev->log, 0, "EINPROGRESS in connect() - selecting");
+                    do {
+                       tv.tv_sec = 15;
+                       tv.tv_usec = 0;
+                       FD_ZERO(&myset);
+                       FD_SET(fd, &myset);
```

Comment:
`select` is a blocking operation, it makes sense using it on the server side to handle multiple sockets. I believe in this case you don't need it, just handle EWOULDBLOCK or check the state of the file descriptor.